### PR TITLE
PB-7476: Adding node affinity to kopia backup and restore jobs

### PR DESCRIPTION
**Pull request by shkumari-px**
**What this PR does / why we need it**: It adds node affinity rule to kopia backup and restore jobs

**Which issue(s) this PR fixes** (optional)
Closes # https://purestorage.atlassian.net/browse/PB-7476

**Special notes for your reviewer**:
The user needs to add this label to kdmp-config **PXB_JOB_NODE_AFFINITY_LABEL:kopia-backup**
This label has to be added to the application node pool as well
**Unit Testing**

1.  Add label to node and kdmp config **Node-2**
Screenshot 2024-08-01 at 10 27 02 AM

2.  Assigned label to a different node, other than node on which the mysql-pod is assigned which we be taking the backup of (mysql app with pvc running on **Node 3**)
mysqlappwithpvcNode

3. Took a backup of default namespace containing mysql app with pvc and noapp-pvc
Job pod created for mysql app with pvc got assigned to Node 3 with no node affinity in it's spec
livebackuppodwithoutnodeaffinity

**Updated result**
Screenshot 2024-08-06 at 1 12 08 PM

Job pod created for noapp-pvc got assigned to node 2 with node affinity in it's spec
nonlivebackupwithnodeafffinity

**Updated result**
Screenshot 2024-08-06 at 1 09 12 PM

**NFS backup Job Node Affinity**
Screenshot 2024-08-01 at 11 06 52 AM
**Updated Result**
Screenshot 2024-08-06 at 1 28 07 PM

**NFS Restore Pod**
Screenshot 2024-07-19 at 11 16 55 AM
**Updated Result**
Screenshot 2024-08-06 at 1 32 30 PM

**Kopia Restore Pod**
Screenshot 2024-07-19 at 11 19 30 AM

**Updated result**
noapp-pvc restore
Screenshot 2024-08-06 at 1 17 32 PM

mysql-pvc restore
Screenshot 2024-08-06 at 1 17 44 PM

csi backup on nfs bl
Screenshot 2024-08-05 at 5 57 13 PM
Screenshot 2024-08-05 at 6 01 49 PM

**nfscsirestore**
Screenshot 2024-08-05 at 6 06 07 PM
Screenshot 2024-08-05 at 6 09 05 PM
**Updated result**
Screenshot 2024-08-06 at 1 51 03 PM

Screenshot 2024-08-06 at 1 51 31 PM
Screenshot 2024-08-06 at 1 52 00 PM

Verified taking backup in psa **environment**
psa environmnet + postgress app with pvc  (backup successful)
psa environment - only pvc (backup failed)
Screenshot 2024-08-06 at 5 59 27 PM
Screenshot 2024-08-06 at 6 02 54 PM
**central ns psa environment**
Screenshot 2024-08-06 at 6 03 51 PM

### CODE REVIEW

File: pkg/controllers/dataexport/reconcile.go [hunk 1]
```diff
@@ -1869,11 +1869,12 @@ func startTransferJob(
...
-		nfsExportPath  string
-		nfsMountOption string
-		psaJobUid      string
-		psaJobGid      string
+		nfsServerAddr              string
+		nfsExportPath              string
+		nfsMountOption             string
+		psaJobUid                  string
+		psaJobGid                  string
+		kopiaBackupJobNodeLabelKey = "KOPIA_BACKUP_NODE_AFFINITY_LABEL"
```
Comment (px-kesavan) [inline pkg/controllers/dataexport/reconcile.go:1877]: can we move this var to pkg/drivers/drivers.go ?

File: pkg/controllers/dataexport/reconcile.go [hunk 2]
```diff
@@ -1885,6 +1886,15 @@ func startTransferJob(
...
 	}
+	nodeLabel := make(map[string]string)
+	kdmpData, err := core.Instance().GetConfigMap(jobConfigMap, jobConfigMapNs)
+	if err != nil {
+		return "", err
+	}
+	kopiaBackupJobNodeLabelValue, ok := kdmpData.Data[kopiaBackupJobNodeLabelKey]
+	if ok && kopiaBackupJobNodeLabelValue != "" {
+		nodeLabel[kopiaBackupJobNodeLabelKey] = kopiaBackupJobNodeLabelValue
+	}
```
Comment (px-kesavan) [inline pkg/controllers/dataexport/reconcile.go:1897]: Can we have a generalized naming ?, as this nodelabel is used for restore as well

File: pkg/controllers/resourceexport/reconcile.go [hunk 3]
```diff
@@ -405,6 +405,16 @@ func startNfsResourceJob(
...
+	var kopiaBackupJobNodeLabelKey = "KOPIA_BACKUP_NODE_AFFINITY_LABEL"
+	nodeLabel := make(map[string]string)
+	kdmpData, err := core.Instance().GetConfigMap(jobConfigMap, jobConfigMapNs)
+	if err != nil {
+		return "", err
+	}
+	kopiaBackupJobNodeLabelValue, ok := kdmpData.Data[kopiaBackupJobNodeLabelKey]
+	if ok && kopiaBackupJobNodeLabelValue != "" {
+		nodeLabel[kopiaBackupJobNodeLabelKey] = kopiaBackupJobNodeLabelValue
+	}
```
Comment (px-kesavan) [inline pkg/controllers/resourceexport/reconcile.go:417]: Same here . Can we have a generalized naming ?, as this nodelabel is used for restore as well

File: pkg/drivers/kopiarestore/kopiarestore.go [hunk 4]
```diff
@@ -306,6 +306,29 @@ func jobFor(
...
+	if len(jobOption.NodeAffinity) > 0 {
+		matchExpressions := []corev1.NodeSelectorRequirement{}
+		for key, val := range jobOption.NodeAffinity {
+			expression := corev1.NodeSelectorRequirement{
+				Key:      key,
+				Operator: corev1.NodeSelectorOpIn,
+				Values:   []string{val},
+			}
+			matchExpressions = append(matchExpressions, expression)
+		}
```
Comment (px-kesavan) [inline pkg/drivers/kopiarestore/kopiarestore.go:318]: same here

File: pkg/drivers/kopiabackup/kopiabackup.go [hunk 5]
```diff
@@ -410,6 +410,31 @@ func jobFor(
...
+					NodeSelectorTerms: []corev1.NodeSelectorTerm{
+						{
+							MatchExpressions: matchExpressions,
+						},
+					},
+				},
+			},
+		}
+	}
+
```
Comment (px-kesavan) [inline pkg/drivers/kopiabackup/kopiabackup.go:437]: @shkumari-px One thing we need test on how this is impacted in RWO(readWriteOnce) PVs. When the pv of RWO access mode is attached to a node (say node 1) and with nodeaffinity we are trying to run the kdmp backup/restore jobs in another node (lets say node 2).

Comment (shkumari-px) [reply pkg/drivers/kopiabackup/kopiabackup.go:437]: > @shkumari-px One thing we need test on how this is impacted in RWO(readWriteOnce) PVs. When the pv of RWO access mode is attached to a node (say node 1) and with nodeaffinity we are trying to run the kdmp backup/restore jobs in another node (lets say node 2).

we get the below error. in this case 
Reason:FailedMount,Message:MountVolume.SetUp failed for volume "pvc-52586bcd-fe1f-4df1-aa70-ba831f096c4c" : rpc error: code = Internal desc = failed to attach volume: Non-shared volume is already attached on another node. Non-shared volumes can only be attached on one node at a time.,Source:EventSource{Component:kubelet,Host:shkumari-116-2,}

here the expectation is to fail the backup
We are adding the node affinity because one of the customer had 2 node pools

- ingress worker node pool (had network restriction)
- application  node pool (application run on this pool only)
here the kdmp jobs got scheduled on ingress worker node pool and the backups were failing

so to fix this we will be adding a label to application node pool and we will be using the same label to schedule the kdmp jobs, this is for non live kdmp jobs

Comment (prashanthpx) [reply pkg/drivers/kopiabackup/kopiabackup.go:437]: @shkumari-px To understand, when kdmp job was tyring to figure out which node was using the app, did it point to ingress worker sometimes and sometime app node pool? Can you add some details here?

Comment (siva-portworx) [reply pkg/drivers/kopiabackup/kopiabackup.go:437]: Actually she  has mentioned the input from the Ranjit in the ticket: https://purestorage.atlassian.net/browse/PB-3647?focusedCommentId=340726
Basically, for non-live PVC backups, customer want the job pods to be schedule on application node pool instead of ingress worker node pool. Since currently we do not have affinity associated to job pods, it lands  on any nodes.
@shkumari-px , @px-kesavan comment is to verify that non-live RWO PVC backup works fine or not.

@px-kesavan , Answer to your question is that PVC in bounded state does not mean that it is attach to any node.  It is just an association to a volume. Only when some pods tries to mount it, it will be get attached to a node and mounted.

```
A PVC (Persistent Volume Claim) being bound means that it has been successfully associated with a Persistent Volume (PV). This does not necessarily mean the volume is attached to a node or mounted in a pod.

Here are the stages involved:

PVC Bound to PV: When a PVC is bound, it means that the Kubernetes control plane has found a suitable PV that matches the PVC’s requirements and has associated them. At this stage, the volume is allocated but not yet attached to any node or mounted by any pod.

Volume Attached to Node: When a pod that references the PVC is scheduled onto a node, the volume is then attached to that node. This means the underlying storage is connected to the node, making it accessible to the node's filesystem.

Volume Mounted in Pod: The final step is when the volume is mounted inside the pod. This happens when the pod is started, and the container(s) within the pod have the specified volume mounted at the designated path within the container's filesystem.

To summarize:

Bound: PVC is matched to a PV.
Attached: Volume is connected to a node.
Mounted: Volume is accessible inside the pod's container.


```

File: pkg/drivers/kopiabackup/kopiabackup.go [hunk 6]
```diff
@@ -519,10 +546,11 @@ func buildJob(jobName string, jobOptions drivers.JobOpts) (*batchv1.Job, error)
 		jobName,
 		resources,
 		nodeName,
+		live,
 	)
 }
 
-func roleFor(live bool) *rbacv1.Role {
+func roleFor() *rbacv1.Role {
```
Comment (vsundarraj-px) [inline pkg/drivers/kopiabackup/kopiabackup.go:553]: why is this removed? Not able to understand the relevance of this change for this fix.

File: pkg/drivers/kopiabackup/kopiabackup.go [hunk 7]
```diff
@@ -532,22 +560,5 @@ func roleFor(live bool) *rbacv1.Role {
 			},
 		},
 	}
-	// Only live backup, we will add the hostaccess and privilege option.
```
Comment (shkumari-px) [inline pkg/drivers/kopiabackup/kopiabackup.go:535]: @vsundarraj-px This code was never used
We had a live bool flag but it was never set to true in the previous already existing code.
Since now we will be setting the live flag to true, this if condition will get triggered now which we don't want
Also this piece of code was unnecessary

Comment (vsundarraj-px) [reply pkg/drivers/kopiabackup/kopiabackup.go:535]: This seems PSA code to me can you confirm this with @lalat-das. @shkumari-px I would prefer a different variable to use here instead of same variable and changing some dead code to avoid confusion. 
If this is a dead code, I think you should remove it as part of another commit not as part of same commit.

Comment (siva-portworx) [reply pkg/drivers/kopiabackup/kopiabackup.go:535]: Or else add a commit message about the cleanup in the same commit as well.

**Comment by siva-portworx** [comment]
Should we also do node affinity for nfscsirestore job as well?

File: pkg/drivers/kopiabackup/kopiabackup.go [hunk 8]
```diff
@@ -410,6 +410,31 @@ func jobFor(
...
+	if len(jobOption.NodeAffinity) > 0 {
+		matchExpressions := []corev1.NodeSelectorRequirement{}
+		for key, val := range jobOption.NodeAffinity {
+			expression := corev1.NodeSelectorRequirement{
+				Key:      key,
+				Operator: corev1.NodeSelectorOpIn,
+				Values:   []string{val},
+			}
+			matchExpressions = append(matchExpressions, expression)
+		}
```
[the remark this answers is not in the document]
Comment (shkumari-px) [reply pkg/drivers/kopiabackup/kopiabackup.go:424]: we can support multiple key-value pairs of node label

File: pkg/drivers/kopiabackup/kopiabackup.go (hunk 6 above)
Comment (shkumari-px) [reply pkg/drivers/kopiabackup/kopiabackup.go:553]: I'll add a separate commit for this cleanup

**Review by prashanthpx — APPROVED** [review]
LGTM

**Review by kgarg-px — APPROVED** [review]

File: pkg/drivers/utils/utils.go [hunk 9]
```diff
@@ -855,6 +855,20 @@ func GetNodeAffinityFromDeployment(name, namespace string) (*corev1.NodeAffinity
 	return deploy.Spec.Template.Spec.Affinity.NodeAffinity, nil
 }
 
+// Get node lablel from deployment
```
Comment (aks-px) [inline pkg/drivers/utils/utils.go:858]: Typo -> label

Comment (aks-px) [reply pkg/drivers/utils/utils.go:858]:
```suggestion
// GetNodeLabelFromDeployment gets node label from deployment
```

**Review by aks-px — APPROVED** [review]
LGTM

**Comment by shkumari-px** [comment]
> Should we also do node affinity for nfscsirestore job as well?

done
Took a backup of thin-csi pvc with thin-csi-sc on nfs backup location
Took a restore of this backup and validated that the node affinity is populated in the restore pod spec
Have attached the screenshot

**Review by siva-portworx — APPROVED** [review]
